pathplannerlib/src/main/native/cpp/pathplanner/lib/PathPlanner.cpp

```diff
@@ -262,19 +257,19 @@ std::vector<PathPlannerTrajectory::EventMarker> PathPlanner::getMarkersFromJson(
 	std::vector < PathPlannerTrajectory::EventMarker > markers;
 
 	if (json.find("markers") != json.end()) {
-		for (wpi::json::reference marker : json.at("markers")) {
+		for (wpi::json::const_reference marker : json.at("markers")) {
 			std::vector < std::string > names;
 			if (marker.find("names") != marker.end()) {
-				for (std::string name : marker.at("names")) {
-					names.push_back(name);
+				for (wpi::json::const_reference name : marker.at("names")) {
+					names.emplace_back(name);
 				}
 			} else {
```

Comment:
This change is causing a build error for me on my machine. Not sure what the difference is but here is the error output:

```
PathPlanner.cpp
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\include\xmemory(676): error C2672: 'std::construct_at': no matching overloaded function found
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\include\vector(602): note: see reference to function template instantiation 'void std::_Default_allocator_traits<_Alloc>::construct<_Ty,const wpi::json::value_type&>(_Alloc &,_Objty *const ,const wpi::json::value_type &)' being compiled
        with
        [
            _Alloc=std::allocator<std::string>,
            _Ty=std::basic_string<char,std::char_traits<char>,std::allocator<char>>,
            _Objty=std::basic_string<char,std::char_traits<char>,std::allocator<char>>
        ]
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\include\vector(607): note: see reference to function template instantiation 'void std::_Default_allocator_traits<_Alloc>::construct<_Ty,const wpi::json::value_type&>(_Alloc &,_Objty *const ,const wpi::json::value_type &)' being compiled
        with
        [
            _Alloc=std::allocator<std::string>,
            _Ty=std::basic_string<char,std::char_traits<char>,std::allocator<char>>,
            _Objty=std::basic_string<char,std::char_traits<char>,std::allocator<char>>
        ]
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\include\vector(620): note: see reference to function template instantiation '_Ty &std::vector<_Ty,std::allocator<_Ty>>::_Emplace_back_with_unused_capacity<const wpi::json::value_type&>(const wpi::json::value_type &)' being compiled
        with
        [
            _Ty=std::string
        ]
C:\Dev\pathplanner\pathplannerlib\src\main\native\cpp\pathplanner\lib\PathPlanner.cpp(264): note: see reference to function template instantiation '_Ty &std::vector<_Ty,std::allocator<_Ty>>::emplace_back<const wpi::json::value_type&>(const wpi::json::value_type &)' being compiled
        with
        [
            _Ty=std::string
        ]
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\include\xmemory(676): error C2783: '_Ty *std::construct_at(_Ty *const ,_Types &&...) noexcept(<expr>)': could not deduce template argument for '<unnamed-symbol>'
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\include\xutility(134): note: see declaration of 'std::construct_at'
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\include\xmemory(680): error C2668: 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>::basic_string': ambiguous call to overloaded function
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\include\xstring(2612): note: could be 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>::basic_string(std::basic_string<char,std::char_traits<char>,std::allocator<char>> &&) noexcept'
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\include\xstring(2960): note: or       'std::basic_string<char,std::char_traits<char>,std::allocator<char>>::basic_string(std::initializer_list<_Elem>,const _Alloc &)'
        with
        [
            _Elem=char,
            _Alloc=std::allocator<char>
        ]
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\include\xmemory(673): note: while trying to match the argument list '(const wpi::json::value_type)'
```

seems to be unable to convert the const_reference to a string.

Reply:
I updated my Visual Studio install and it still fails. I have no idea why it doesn't fail on your end.

Reply:
Your compiler doesn't seem to enjoy solving puzzles. I'll rewrite to be a little bit more explicit.

Reply:
Hopefully the new commit works

Reply:
Getting the same error unfortunately. Is your install using the MSVC compiler as well or something different?

Reply:
It doesn't fail for me building for the RIO, as it does when running CI, so that would make sense. Although it should be building desktop for you as well unless you're explicitly telling it not to.

Reply:
That would make sense. I didn't get to it last night, but I'll attempt to write a fix that works for Visual Studio as well this evening.

Reply:
I have visual studio installed now, and I get the same issue. However, it remains even when testing code from an older commit, unless I missed something. I'll do more testing later in the weekend

Reply:
Does the main branch build? the main branch builds fine for me.

Reply:
Yes, the main branch builds fine. The build was still failing despite me reverting the lines highlighted, because a similar issue arose on line 268. I fixed both by replacing emplace_back with push_back. It seems MSVC and gcc have different implementations for dealing with emplace_back and push_back (gcc simply calls emplace_back, but MSVC seems to handle this differently). Also, the wpi::json class is written in a very interesting way, because they handle types themselves as an enum `value_t` instead of as a template_argument. I'm sure this has its benefits, but it makes determining types a nightmare for the compiler. I'm pleasantly surprised that gcc can pull it off, but MSVC seems to be a little lacking, and gives up when it comes to using emplace_back for some reason. Anyways, for now, this builds.

Reply:
It's also possible MSVC handles the construction of std::string differently, since it's complaining about not being able to determine whether to use const& or && (copy or move).

Reply:
If you force it to use const & like this, it compiles just fine

`				
for (wpi::json::const_reference name : marker.at("names")) {
std::string const &ref = name;
names.emplace_back(name);
}
`

Reply:
But for now I think just using what I committed should work best.